File: go/indicators/custom/maximumentropyspectrum/maximumentropyspectrum.go
// Package maximumentropyspectrum implements MBST's Maximum Entropy Spectrum heatmap indicator.
//
// The Maximum Entropy Spectrum (MESPECT) displays a power heatmap of the cyclic
// activity over a configurable cycle-period range using Burg's maximum-entropy
// auto-regressive method. It supports a fast-attack slow-decay automatic gain
// control and either floating or fixed (0-clamped) intensity normalization.
//
// Reference: MBST Mbs.Trading.Indicators.SpectralAnalysis.MaximumEntropySpectrum.
package maximumentropyspectrum

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// MaximumEntropySpectrum is MBST's Maximum Entropy Spectrum heatmap indicator.
type MaximumEntropySpectrum struct {
	mu                    sync.RWMutex
	mnemonic              string
	description           string
	estimator             *estimator
	windowCount           int
	lastIndex             int
	primed                bool
	floatingNormalization bool
	minParameterValue     float64
	maxParameterValue     float64
	parameterResolution   float64
	barFunc               entities.BarFunc
	quoteFunc             entities.QuoteFunc
	tradeFunc             entities.TradeFunc
}

// NewMaximumEntropySpectrumDefault returns an instance created with default parameters.
func NewMaximumEntropySpectrumDefault() (*MaximumEntropySpectrum, error) {
	return NewMaximumEntropySpectrumParams(&Params{})
}

// NewMaximumEntropySpectrumParams returns an instance created with the supplied parameters.
//
//nolint:funlen,cyclop
func NewMaximumEntropySpectrumParams(p *Params) (*MaximumEntropySpectrum, error) {
	const (
		invalid           = "invalid maximum entropy spectrum parameters"
		fmtLength         = "%s: Length should be >= 2"
		fmtDegree         = "%s: Degree should be > 0 and < Length"
		fmtMinPeriod      = "%s: MinPeriod should be >= 2"
		fmtMaxPeriod      = "%s: MaxPeriod should be > MinPeriod"
		fmtNyquist        = "%s: MaxPeriod should be <= 2 * Length"
		fmtResolution     = "%s: SpectrumResolution should be >= 1"
		fmtAgc            = "%s: AutomaticGainControlDecayFactor should be in (0, 1)"
		fmtw              = "%s: %w"
		descrPrefix       = "Maximum entropy spectrum "
		defLength         = 60
		defDegree         = 30
		defMinPeriod      = 2.0
		defMaxPeriod      = 59.0
		defSpectrumResult = 1
		defAgcDecayFactor = 0.995
		agcDecayEpsilon   = 1e-12
	)

	cfg := *p

	if cfg.Length == 0 {
		cfg.Length = defLength
	}

	if cfg.Degree == 0 {
		cfg.Degree = defDegree
	}

	if cfg.MinPeriod == 0 {
		cfg.MinPeriod = defMinPeriod
	}

	if cfg.MaxPeriod == 0 {
		cfg.MaxPeriod = defMaxPeriod
	}

	if cfg.SpectrumResolution == 0 {
		cfg.SpectrumResolution = defSpectrumResult
	}

	if cfg.AutomaticGainControlDecayFactor == 0 {
		cfg.AutomaticGainControlDecayFactor = defAgcDecayFactor
	}

	agcOn := !cfg.DisableAutomaticGainControl
	floatingNorm := !cfg.FixedNormalization

	if cfg.Length < 2 {
		return nil, fmt.Errorf(fmtLength, invalid)
	}

	if cfg.Degree <= 0 || cfg.Degree >= cfg.Length {
		return nil, fmt.Errorf(fmtDegree, invalid)
	}

	if cfg.MinPeriod < 2 {
		return nil, fmt.Errorf(fmtMinPeriod, invalid)
	}

	if cfg.MaxPeriod <= cfg.MinPeriod {
		return nil, fmt.Errorf(fmtMaxPeriod, invalid)
	}

	if cfg.MaxPeriod > 2*float64(cfg.Length) {
		return nil, fmt.Errorf(fmtNyquist, invalid)
	}

	if cfg.SpectrumResolution < 1 {
		return nil, fmt.Errorf(fmtResolution, invalid)
	}

	if agcOn &&
		(cfg.AutomaticGainControlDecayFactor <= 0 || cfg.AutomaticGainControlDecayFactor >= 1) {
		return nil, fmt.Errorf(fmtAgc, invalid)
	}

	// MaximumEntropySpectrum mirrors MBST's reference: BarMedianPrice default.
	bc := cfg.BarComponent
	if bc == 0 {
		bc = entities.BarMedianPrice
	}

	qc := cfg.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := cfg.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	barFunc, err := entities.BarComponentFunc(bc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	quoteFunc, err := entities.QuoteComponentFunc(qc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	tradeFunc, err := entities.TradeComponentFunc(tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)

	flags := buildFlagTags(&cfg, agcOn, floatingNorm, defAgcDecayFactor, agcDecayEpsilon)
	mnemonic := fmt.Sprintf("mespect(%d, %d, %g, %g, %d%s%s)",
		cfg.Length, cfg.Degree, cfg.MinPeriod, cfg.MaxPeriod, cfg.SpectrumResolution,
		flags, componentMnemonic)

	est := newEstimator(
		cfg.Length, cfg.Degree, cfg.MinPeriod, cfg.MaxPeriod, cfg.SpectrumResolution,
		agcOn, cfg.AutomaticGainControlDecayFactor,
	)

	return &MaximumEntropySpectrum{
		mnemonic:              mnemonic,
		description:           descrPrefix + mnemonic,
		estimator:             est,
		lastIndex:             cfg.Length - 1,
		floatingNormalization: floatingNorm,
		minParameterValue:     cfg.MinPeriod,
		maxParameterValue:     cfg.MaxPeriod,
		parameterResolution:   float64(cfg.SpectrumResolution),
		barFunc:               barFunc,
		quoteFunc:             quoteFunc,
		tradeFunc:             tradeFunc,
	}, nil
}

// buildFlagTags encodes non-default boolean/decay settings as terse override-only tags.
// Returns an empty string when all flags are at their defaults.
func buildFlagTags(cfg *Params, agcOn, floatingNorm bool, defAgc, eps float64) string {
	var s string

	if !agcOn {
		s += ", no-agc"
	}

	if agcOn && math.Abs(cfg.AutomaticGainControlDecayFactor-defAgc) > eps {
		s += fmt.Sprintf(", agc=%g", cfg.AutomaticGainControlDecayFactor)
	}

	if !floatingNorm {
		s += ", no-fn"
	}

	return s
}

// IsPrimed indicates whether the indicator is primed.
func (s *MaximumEntropySpectrum) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *MaximumEntropySpectrum) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.MaximumEntropySpectrum,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: s.description},
		},
	)
}

// Update feeds the next sample to the engine and returns the heatmap column.
//
// Before priming the heatmap is empty (with the indicator's parameter axis).
// On a NaN input sample the state is left unchanged and an empty heatmap is
// returned.
func (s *MaximumEntropySpectrum) Update(sample float64, t time.Time) *outputs.Heatmap {
	s.mu.Lock()
	defer s.mu.Unlock()

	if math.IsNaN(sample) {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution)
	}

	window := s.estimator.inputSeries

	if s.primed {
		copy(window[:s.lastIndex], window[1:])
		window[s.lastIndex] = sample
	} else {
		window[s.windowCount] = sample
		s.windowCount++

		if s.windowCount == s.estimator.length {
			s.primed = true
		}
	}

	if !s.primed {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution)
	}

	s.estimator.calculate()

	lengthSpectrum := s.estimator.lengthSpectrum

	var minRef float64
	if s.floatingNormalization {
		minRef = s.estimator.spectrumMin
	}

	maxRef := s.estimator.spectrumMax
	spectrumRange := maxRef - minRef

	// MBST fills spectrum[0] at MaxPeriod and spectrum[last] at MinPeriod.
	// The heatmap axis runs MinPeriod -> MaxPeriod, so reverse on output.
	values := make([]float64, lengthSpectrum)
	valueMin := math.Inf(1)
	valueMax := math.Inf(-1)

	for i := 0; i < lengthSpectrum; i++ {
		v := (s.estimator.spectrum[lengthSpectrum-1-i] - minRef) / spectrumRange
		values[i] = v

		if v < valueMin {
			valueMin = v
		}

		if v > valueMax {
			valueMax = v
		}
	}

	return outputs.NewHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution,
		valueMin, valueMax, values)
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *MaximumEntropySpectrum) UpdateScalar(sample *entities.Scalar) core.Output {
	return s.updateEntity(sample.Time, sample.Value)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *MaximumEntropySpectrum) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, s.barFunc(sample))
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *MaximumEntropySpectrum) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, s.quoteFunc(sample))
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *MaximumEntropySpectrum) UpdateTrade(sample *entities.Trade) core.Output {
	return s.updateEntity(sample.Time, s.tradeFunc(sample))
}

func (s *MaximumEntropySpectrum) updateEntity(t time.Time, sample float64) core.Output {
	const length = 1

	heatmap := s.Update(sample, t)

	output := make([]any, length)
	output[0] = heatmap

	return output
}

File: ts/indicators/custom/maximum-entropy-spectrum/maximum-entropy-spectrum.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { Heatmap } from '../../core/outputs/heatmap';
import { MaximumEntropySpectrumEstimator } from './estimator';
import { MaximumEntropySpectrumParams } from './params';

const DEFAULT_LENGTH = 60;
const DEFAULT_DEGREE = 30;
const DEFAULT_MIN_PERIOD = 2;
const DEFAULT_MAX_PERIOD = 59;
const DEFAULT_SPECTRUM_RESOLUTION = 1;
const DEFAULT_AGC_DECAY_FACTOR = 0.995;
const AGC_DECAY_EPSILON = 1e-12;

/** __Maximum Entropy Spectrum__ heatmap indicator (MBST port).
 *
 * Displays a power heatmap of cyclic activity over a configurable cycle-period range using
 * Burg's maximum-entropy auto-regressive method. It supports a fast-attack slow-decay
 * automatic gain control and either floating or fixed (0-clamped) intensity normalization.
 *
 * Reference: MBST Mbs.Trading.Indicators.SpectralAnalysis.MaximumEntropySpectrum. */
export class MaximumEntropySpectrum implements Indicator {
  private readonly estimator: MaximumEntropySpectrumEstimator;
  private readonly lastIndex: number;
  private readonly floatingNormalization: boolean;
  private readonly minParameterValue: number;
  private readonly maxParameterValue: number;
  private readonly parameterResolution: number;

  private readonly mnemonicValue: string;
  private readonly descriptionValue: string;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  private windowCount = 0;
  private primed = false;

  /** Creates an instance with default parameters
   * (length=60, degree=30, minPeriod=2, maxPeriod=59, spectrumResolution=1,
   * AGC on (decay=0.995), floating normalization, BarComponent.Median). */
  public static default(): MaximumEntropySpectrum {
    return new MaximumEntropySpectrum({});
  }

  /** Creates an instance based on the given parameters. */
  public static fromParams(params: MaximumEntropySpectrumParams): MaximumEntropySpectrum {
    return new MaximumEntropySpectrum(params);
  }

  private constructor(params: MaximumEntropySpectrumParams) {
    const invalid = 'invalid maximum entropy spectrum parameters';

    const length = params.length !== undefined && params.length !== 0
      ? params.length : DEFAULT_LENGTH;
    const degree = params.degree !== undefined && params.degree !== 0
      ? params.degree : DEFAULT_DEGREE;
    const minPeriod = params.minPeriod !== undefined && params.minPeriod !== 0
      ? params.minPeriod : DEFAULT_MIN_PERIOD;
    const maxPeriod = params.maxPeriod !== undefined && params.maxPeriod !== 0
      ? params.maxPeriod : DEFAULT_MAX_PERIOD;
    const spectrumResolution = params.spectrumResolution !== undefined && params.spectrumResolution !== 0
      ? params.spectrumResolution : DEFAULT_SPECTRUM_RESOLUTION;
    const agcDecayFactor = params.automaticGainControlDecayFactor !== undefined
      && params.automaticGainControlDecayFactor !== 0
      ? params.automaticGainControlDecayFactor : DEFAULT_AGC_DECAY_FACTOR;

    const agcOn = !params.disableAutomaticGainControl;
    const floatingNorm = !params.fixedNormalization;

    if (length < 2) {
      throw new Error(`${invalid}: Length should be >= 2`);
    }
    if (degree <= 0 || degree >= length) {
      throw new Error(`${invalid}: Degree should be > 0 and < Length`);
    }
    if (minPeriod < 2) {
      throw new Error(`${invalid}: MinPeriod should be >= 2`);
    }
    if (maxPeriod <= minPeriod) {
      throw new Error(`${invalid}: MaxPeriod should be > MinPeriod`);
    }
    if (maxPeriod > 2 * length) {
      throw new Error(`${invalid}: MaxPeriod should be <= 2 * Length`);
    }
    if (spectrumResolution < 1) {
      throw new Error(`${invalid}: SpectrumResolution should be >= 1`);
    }
    if (agcOn && (agcDecayFactor <= 0 || agcDecayFactor >= 1)) {
      throw new Error(`${invalid}: AutomaticGainControlDecayFactor should be in (0, 1)`);
    }

    // MaximumEntropySpectrum mirrors MBST's reference: BarComponent.Median default.
    const bc = params.barComponent ?? BarComponent.Median;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    this.estimator = new MaximumEntropySpectrumEstimator(
      length, degree, minPeriod, maxPeriod, spectrumResolution,
      agcOn, agcDecayFactor,
    );
    this.lastIndex = length - 1;
    this.floatingNormalization = floatingNorm;
    this.minParameterValue = minPeriod;
    this.maxParameterValue = maxPeriod;
    this.parameterResolution = spectrumResolution;

    const cm = componentTripleMnemonic(bc, qc, tc);
    const flags = buildFlagTags(agcOn, floatingNorm, agcDecayFactor);
    this.mnemonicValue = `mespect(${length}, ${degree}, ${formatNum(minPeriod)}, ${formatNum(maxPeriod)}, ${spectrumResolution}${flags}${cm})`;
    this.descriptionValue = 'Maximum entropy spectrum ' + this.mnemonicValue;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean { return this.primed; }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.MaximumEntropySpectrum,
      this.mnemonicValue,
      this.descriptionValue,
      [
        { mnemonic: this.mnemonicValue, description: this.descriptionValue },
      ],
    );
  }

  /** Feeds the next sample to the engine and returns the heatmap column.
   *
   * Before priming the heatmap is empty (with the indicator's parameter axis).
   * On a NaN input sample the state is left unchanged and an empty heatmap is returned. */
  public update(sample: number, time: Date): Heatmap {
    if (Number.isNaN(sample)) {
      return Heatmap.newEmptyHeatmap(time, this.minParameterValue, this.maxParameterValue, this.parameterResolution);
    }

    const window = this.estimator.inputSeries;

    if (this.primed) {
      for (let i = 0; i < this.lastIndex; i++) {
        window[i] = window[i + 1];
      }
      window[this.lastIndex] = sample;
    } else {
      window[this.windowCount] = sample;
      this.windowCount++;
      if (this.windowCount === this.estimator.length) {
        this.primed = true;
      }
    }

    if (!this.primed) {
      return Heatmap.newEmptyHeatmap(time, this.minParameterValue, this.maxParameterValue, this.parameterResolution);
    }

    this.estimator.calculate();

    const lengthSpectrum = this.estimator.lengthSpectrum;

    const minRef = this.floatingNormalization ? this.estimator.spectrumMin : 0;
    const maxRef = this.estimator.spectrumMax;
    const spectrumRange = maxRef - minRef;

    // MBST fills spectrum[0] at MaxPeriod and spectrum[last] at MinPeriod.
    // The heatmap axis runs MinPeriod -> MaxPeriod, so reverse on output.
    const values = new Array<number>(lengthSpectrum);
    let valueMin = Number.POSITIVE_INFINITY;
    let valueMax = Number.NEGATIVE_INFINITY;

    for (let i = 0; i < lengthSpectrum; i++) {
      const v = (this.estimator.spectrum[lengthSpectrum - 1 - i] - minRef) / spectrumRange;
      values[i] = v;
      if (v < valueMin) valueMin = v;
      if (v > valueMax) valueMax = v;
    }

    return Heatmap.newHeatmap(
      time, this.minParameterValue, this.maxParameterValue, this.parameterResolution,
      valueMin, valueMax, values,
    );
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    return this.updateEntity(sample.time, sample.value);
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    return this.updateEntity(sample.time, this.barComponentFunc(sample));
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    return this.updateEntity(sample.time, this.quoteComponentFunc(sample));
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    return this.updateEntity(sample.time, this.tradeComponentFunc(sample));
  }

  private updateEntity(time: Date, sample: number): IndicatorOutput {
    return [this.update(sample, time)];
  }
}

/** Encodes non-default boolean/decay settings as terse override-only tags. Empty when all
 * flags are at their defaults. */
function buildFlagTags(agcOn: boolean, floatingNorm: boolean, agcDecayFactor: number): string {
  let s = '';
  if (!agcOn) s += ', no-agc';
  if (agcOn && Math.abs(agcDecayFactor - DEFAULT_AGC_DECAY_FACTOR) > AGC_DECAY_EPSILON) {
    s += `, agc=${formatNum(agcDecayFactor)}`;
  }
  if (!floatingNorm) s += ', no-fn';
  return s;
}

/** Matches Go fmt.Sprintf("%g") for the common integer and decimal cases used in the mnemonic. */
function formatNum(n: number): string {
  return n.toString();
}
